.github/workflows/CICD.yaml

```diff
@@ -8,53 +8,264 @@ on:
+
+          echo 'prefix=${{ github.workspace }}/node_modules' >> ~/.npmrc
+          # Note that currently pyproject.toml does not specify a version.
+          # As such, this will install the latest Pyright, including
+          # potentionally breaking changes.
+          npm install -g pyright
+          ./install_all.sh -e
+
+      # TODO see note above about 3rd party dependencies.
+      #- name: Save dependency cache
+      #  if: ${{ success() && !steps.restore-dependency-cache.outputs.cache-hit }}
+      #  uses: actions/cache/save@v4
+      #  id: save-dependency-cache
+      #  with:
+      #    path: |
+      #      .github-venv
+      #      node_modules
+      #    key: ${{ steps.generate-cache-keys.outputs.cache_key_dependencies }}
+
+      - name: Save run cache
+        uses: actions/cache/save@v4
+        id: save-run-cache
+        with:
+          path: ${{ github.workspace }}
+          key: ${{ steps.generate-cache-keys.outputs.cache_key_run }}
+
+  Pyright:
+    name: Static type checking
+    runs-on: ubuntu-latest
+    needs:
+      - Checkout
+
+    env:
+      PYTHON_VERSION: "3.10"
+
+    if: ${{( success() && !cancelled() ) }}
+
+    steps:
+      - name: Set up Python ${{ env.PYTHON_VERSION }}
+        uses: actions/setup-python@v5
+        id: install_python
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+
+      - uses: actions/cache/restore@v4
+        name: Restore run cache
+        id: restore-run-cache
+        with:
+          key: ${{ needs.Checkout.outputs.cache-key-run }}
+          path: ${{ github.workspace }}
+          fail-on-cache-miss: true
 
       - name: Test with pyright
         run: |
-          pyright
+          echo Running pyright
+          source ${{ github.workspace }}/.github-venv/bin/activate
+          ./node_modules/bin/pyright
+
+  Pytest:
+    name: Automated testing
+    runs-on: ubuntu-latest
+    needs:
+      - Checkout
+
+    env:
+      PYTHON_VERSION: "3.10"
+
+    if: ${{( success() && !cancelled() ) }}
 
-      - name: Test with pytest
+    steps:
+      - name: Set up Python ${{ env.PYTHON_VERSION }}
+        uses: actions/setup-python@v5
+        id: install_python
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+
+      - uses: actions/cache/restore@v4
+        name: Restore run cache
+        id: restore-run-cache
+        with:
+          key: ${{ needs.Checkout.outputs.cache-key-run }}
+          path: ${{ github.workspace }}
+          fail-on-cache-miss: true
+
+      - name: Test with pytest and generate reports
         run: |
           echo Running pytest
-          # infra is not set up to support integration tests, so ignore postgresql
-          # etl has been updated, while the tests have not been
-          pytest --junitxml pytest.xml \
-                 -k "not acceptance"
+          source ${{ github.workspace }}/.github-venv/bin/activate
+          pytest -k "not acceptance"
```

Comment:
A note while we're in here:

This `-k not acceptance` is from CAP. I just added this as a way to (primitively) exclude acceptance tests from the first run of unit/integration tests during CICD. We might want to remove or generalize this so BLPython users are free to structure their tests however they need to. I don't think this affects these changes, though.

Reply:
Indeed, I left it here for the same reason we don't want to run acceptance tests for CAP - that is, they tend to require resources beyond what we have in GitHub Actions. While there are not currently any acceptance tests, I want to follow the same established testing patterns we've been using for CAP wherever it makes sense.

Reply:
Right, totally agree on the intent. Just thinking that `-k not acceptance` may not work for every BLPython application if someone structures their test names or directories differently. We can cross this bridge later!

Reply:
A goal of the monorepo approach is that all packages will follow the same principles and guidelines. While it's more restrictive, it means fewer development preferences to maintain and support. So yes, we will certainly defer questions around how to name tests, but enforcing whatever decision takes place will be applied consistently across this repository.